[FLINK-17331] Explicitly get the ByteBuf length of all classes which …

KarmaGYZ opened the pull request:
…is written to NettyMessage

Currently, the length of some header fields in NettyMessage is hardcoded: InputChannelID, ExecutionAttemptID, e.t.c.
So if we make some changes for such field, then we are not ware that it also needs to change the respective length for related netty messages component.
This PR explicitly get the ByteBuf length of all classes which is written to NettyMessage to avoid such problems.

<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

Explicitly get the ByteBuf length of all classes which is written to NettyMessage

KarmaGYZ commented:
cc @zhijiangW

KarmaGYZ commented:
cc @curcur Rebased to the latest master branch.

curcur commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/InputChannelID.java:
```diff
@@ -28,6 +28,7 @@
 public class InputChannelID extends AbstractID {
 
 	private static final long serialVersionUID = 1L;
+	private static final int BYTEBUF_LEN = 16;
```
similarly here, please add a comment to explain.

curcur commented on flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/IntermediateResultPartitionID.java:
```diff
@@ -30,6 +30,7 @@
 public class IntermediateResultPartitionID implements ResultID {
 
 	private static final long serialVersionUID = 1L;
+	private static final int BYTEBUF_LEN = 20;
```
similarly here, please add a comment. 
Where are the extra 4 bytes coming from? partitionNum?

KarmaGYZ replied:
Yes, the extra 4 bytes come from the `partitionNum`.

curcur commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/NettyMessage.java:
```diff
@@ -259,7 +259,7 @@ protected Object decode(ChannelHandlerContext ctx, ByteBuf in) throws Exception
 
 		// receiver ID (16), sequence number (4), backlog (4), dataType (1), isCompressed (1), buffer size (4)
-		static final int MESSAGE_HEADER_LENGTH = 16 + 4 + 4 + 1 + 1 + 4;
+		static final int MESSAGE_HEADER_LENGTH = InputChannelID.getByteBufLength() + Integer.BYTES + Integer.BYTES + Byte.BYTES + 1 + Integer.BYTES;
```
why `isCompressed` (boolean) is hardcoded, while everything else is not?

KarmaGYZ replied:
After a deeper investigation, I find that a `boolean` will be treated as a `byte` in netty. Will change it to `Byte.BYTES`.

curcur commented:
@KarmaGYZ , thanks for the PR.

I think the PR is quite safe to go, and I left a few minor comments.

Besides that, I have a few comments on the filed PR itself.

1. Maybe add a bit more detail the motivation of this change in the section of "What is the purpose of the change".
2. Would you please add a test for NettyMessage.java as well? No matter what, that's the main purpose of the change.

KarmaGYZ commented:
Thanks for the review, @curcur !
I've addressed your comments. Regarding the test, I think `NettyMessageServerSideSerializationTest` might be good enough as it covers the encoding and decoding logic for all the `NettyMessage`s.

curcur commented:
> Thanks for the review, @curcur !
> I've addressed your comments. Regarding the test, I think `NettyMessageServerSideSerializationTest` might be good enough as it covers the encoding and decoding logic for all the `NettyMessage`s.

That' true, thanks for pointing.

curcur commented:
Just one more thing: I think the description could be a bit more informative: something like this:

"""
Currently, the length of some header fields in NettyMessage is hardcoded: InputChannelID, ExecutionAttemptID, e.t.c.
So if we make some changes for such field, then we are not ware that it also needs to change the respective length for related netty messages component.
This PR explicitly get the ByteBuf length of all classes which is written to NettyMessage to avoid such problems.
"""

pnowojski reviewed (commented):
Thanks for the contribution @KarmaGYZ  and review @curcur.

Test failure is unrelated and was fixed on the master https://issues.apache.org/jira/browse/FLINK-20793, so merging
